File: ClientMessageForwarder.js
function ClientMessageForwarder(redis_connection, amqpConnection) {
    this.EVENT_EXCHANGE = "openspy.master";
    this.EVENT_ROUTEKEY = "qr.message";
    this.EVENT_CLIENT_MESSAGE_ACK_ROUTEKEY = "client-messages.acks";
    this.EVENT_SB_MESSAGE_ROUTEKEY = "client.message";

    this.amqpConnection = amqpConnection;
    this.redis_connection = redis_connection;

    this.client_message_resend_timers = {};
    this.CLIENT_MESSAGE_RESNED_INTERVAL = 1000;
    this.CLIENT_MESSAGE_MAX_RESENDS = 5;

    this.amqpConnection.createChannel(function(err, ch) {
        if(err) {
            console.error(err);
            process.exit(-1);
        }
        ch.on('error', (err) => {
            console.error(err);
            process.exit(-1);
        });
        this.channel = ch;
        ch.assertExchange(this.EVENT_EXCHANGE, 'topic', {durable: true});
        ch.assertQueue('', {exclusive: true}, function(err, q) {
            ch.bindQueue(q.queue, this.EVENT_EXCHANGE, this.EVENT_CLIENT_MESSAGE_ACK_ROUTEKEY);

            ch.consume(q.queue, this.OnClientMessageAck.bind(this), {noAck: true});
        }.bind(this));

        ch.assertQueue('', {exclusive: true}, function(err, q) {
            ch.bindQueue(q.queue, this.EVENT_EXCHANGE, this.EVENT_SB_MESSAGE_ROUTEKEY);

            ch.consume(q.queue, this.OnGotClientMessageFromServerBrowsing.bind(this), {noAck: true});
        }.bind(this));  
        
    }.bind(this));
}

ClientMessageForwarder.prototype.GetClientMessageUniqueKey = function(driver_hostname, driver_address, instance_key, to_address, key) {
    return driver_hostname + ":" + driver_address + ":" + to_address + ":" + instance_key + ":" + key;
}
ClientMessageForwarder.prototype.OnGotClientMessageFromServerBrowsing = function(message) {
    var sb_forwarded_message = message.content.toString();
    var msg_split = sb_forwarded_message.split("\\");        

    var to_ip = msg_split[5]
    var to_port = msg_split[6];
    var message = msg_split[7];
    var ip_string = to_ip + ":" + to_port;
    this.ForwardMessageToQR(ip_string, message);   
}
ClientMessageForwarder.prototype.OnClientMessageAck = function(message) {
    var msg = JSON.parse(message.content.toString());
    var key = this.GetClientMessageUniqueKey(msg.hostname, msg.driver_address, msg.instance_key, msg.from_address, msg.identifier);
    if(this.client_message_resend_timers[key] !== undefined) {
        clearInterval(this.client_message_resend_timers[key].timer);
        delete this.client_message_resend_timers[key];
    }
}

ClientMessageForwarder.prototype.ForwardMessageToQR = function(to_address, b64_string) {
    var key = "IPMAP_" + to_address.replace(":", "-");
    this.redis_connection.get(key,  function(err, res) {
        if(err) throw err;
        if(res != null) {
            this.redis_connection.hmget(res, ["instance_key", "driver_hostname", "driver_address"], function(err, res) {

                if(!res || !res.length || !res[0]) return;

                //generate unique key, and add to resend queue
                //forward to QR every 1 second, up to 5 times - or until unique key ack is received
                var identifier = Math.floor(Math.random() * ((Math.pow(2,32)/2)-1));
                var event_message = {hostname: res[1], driver_address: res[2], to_address, message: b64_string, version: 2, instance_key: parseInt(res[0]), identifier, type: "client_message"};
                var interval_key = this.GetClientMessageUniqueKey(event_message.hostname, event_message.driver_address, event_message.instance_key, event_message.to_address, event_message.identifier);
                if(this.client_message_resend_timers[interval_key] === undefined) {
                    this.client_message_resend_timers[interval_key] = {resends: 0};
                } else {
                    return; //XXX: this should not happen, throw error
                }
                this.client_message_resend_timers[interval_key].timer = setInterval(function(resend_data, interval_key, event_message) {
                    if(resend_data.resends > this.CLIENT_MESSAGE_MAX_RESENDS) {
                        clearInterval(resend_data.timer);
                        delete this.client_message_resend_timers[interval_key];
                        return;
                    }
                    resend_data.resends++;                    
                    var msg = Buffer.from(JSON.stringify(event_message));
                    this.channel.publish(this.EVENT_EXCHANGE, this.EVENT_ROUTEKEY, msg);
                }.bind(this, this.client_message_resend_timers[interval_key], interval_key, event_message), this.CLIENT_MESSAGE_RESNED_INTERVAL)
            }.bind(this));
        }

    }.bind(this));


}
module.exports = ClientMessageForwarder;

File: src/Handlers/ClientMessageForwarder.go
package Handlers

import (
	"context"
	"encoding/json"
	"log"
	"math/rand/v2"
	"net/netip"
	"os-qr-service/Helpers"
	"os-qr-service/Server"
	"strconv"
	"strings"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/redis/go-redis/v9"
)

const (
	QR_MESSAGE_ROUTE_KEY            string = "client.message"
	QR_CLIENT_MESSAGE_ACK_ROUTE_KEY        = "client-messages.acks"
)

const NUM_CLIENT_MESSAGE_RETRIES int = 3

type QRMessage struct {
	Hostname      string `json:"hostname"`
	DriverAddress string `json:"driver_address"`
	ToAddress     string `json:"to_address"`
	FromAddress   string `json:"from_address"` //this is inverted in the ack..
	Message       string `json:"message"`
	Version       int    `json:"version"`
	InstanceKey   int    `json:"instance_key"`
	Identifier    int32  `json:"identifier"`
	Type          string `json:"type"`
	Retries       int
	LastResend    time.Time
}

type ClientMessageForwarder struct {
	context        context.Context
	redisOptions   *redis.Options
	redisClient    *redis.Client
	amqpChannel    *amqp.Channel
	amqpQueue      amqp.Queue
	resendTicker   *time.Ticker
	serverManager  Server.IServerManager
	serverGroupMgr Server.IServerGroupManager
	gameManager    Server.IGameManager

	forwardedMessages map[string]QRMessage
}

func (h *ClientMessageForwarder) SetManagers(redisOptions *redis.Options, context context.Context, serverMgr Server.IServerManager, serverGroupMgr Server.IServerGroupManager, gameMgr Server.IGameManager) {
	h.redisOptions = redisOptions
	h.redisOptions.DB = 0
	h.redisClient = redis.NewClient(h.redisOptions)
	h.context = context
	h.serverManager = serverMgr
	h.forwardedMessages = make(map[string]QRMessage)
}

func (h *ClientMessageForwarder) HandleNewServer(serverKey string) {
}
func (h *ClientMessageForwarder) HandleUpdateServer(serverKey string) {

}
func (h *ClientMessageForwarder) HandleDeleteServer(serverKey string) {

}

func (h *ClientMessageForwarder) SetAMQPConnection(amqpConnection *amqp.Connection) {
	chListen, err := amqpConnection.Channel()
	if err != nil {
		Helpers.FailOnError(err, "Failed to open client message forwarder channel")
	}
	h.amqpChannel = chListen

	q, err := h.amqpChannel.QueueDeclare(
		"qr-service-clientmsg", // name
		false,                  // durable
		true,                   // delete when unused
		true,                   // exclusive
		false,                  // no-wait
		nil,                    // arguments
	)
	Helpers.FailOnError(err, "Failed to declare client message forwarder queue")
	h.amqpChannel.QueueBind(q.Name, QR_MESSAGE_ROUTE_KEY, "openspy.master", false, nil)
	h.amqpChannel.QueueBind(q.Name, QR_CLIENT_MESSAGE_ACK_ROUTE_KEY, "openspy.master", false, nil)
	h.amqpQueue = q
	h.resendTicker = time.NewTicker(1 * time.Second)

	go h.listenLoop()
}

func (h *ClientMessageForwarder) listenLoop() {
	msgs, err := h.amqpChannel.Consume(
		h.amqpQueue.Name, // queue
		"",               // consumer
		true,             // auto-ack
		true,             // exclusive
		false,            // no-local
		false,            // no-wait
		nil,              // args
	)
	Helpers.FailOnError(err, "Failed to consume client message forwarder queue")
	for {
		select {
		case d := <-msgs:
			if d.RoutingKey == QR_MESSAGE_ROUTE_KEY {
				h.OnGotClientMessageFromServerBrowsing(d)
			} else if d.RoutingKey == QR_CLIENT_MESSAGE_ACK_ROUTE_KEY {
				h.OnClientMessageAck(d)
			}
		case <-h.resendTicker.C:
			h.submitResends()
		}
	}

}
func (h *ClientMessageForwarder) OnGotClientMessageFromServerBrowsing(amqpMsg amqp.Delivery) {
	//forward to QR
	var msg = string(amqpMsg.Body)

	var splitMsg = strings.Split(msg, "\\")
	log.Println(splitMsg)

	var toIp = splitMsg[5]
	var toPort = splitMsg[6]
	var message = splitMsg[7]

	addrPort, err := netip.ParseAddrPort(toIp + ":" + toPort)
	if err != nil && len(err.Error()) > 0 {
		log.Printf("OnGotClientMessageFromServerBrowsing ip parse error: %s\n", err.Error())
		return
	}
	h.ForwardMessageToQR(addrPort, message)
}

func (h *ClientMessageForwarder) ForwardMessageToQR(addrPort netip.AddrPort, base64str string) {
	var serverKey = h.serverManager.GetServerKeyFromAddress(h.context, h.redisClient, addrPort)
	if len(serverKey) == 0 {
		return
	}
	var results = h.serverManager.GetKeys(h.context, h.redisClient, serverKey, "instance_key", "driver_hostname", "driver_address")

	instanceKey, parseError := strconv.Atoi(results[0])
	if parseError != nil && len(parseError.Error()) > 0 {
		log.Printf("ForwardMessageToQR instance key parse error: %s\n", parseError.Error())
		return
	}

	var qrMsg QRMessage
	qrMsg.Message = base64str
	qrMsg.Version = 2
	qrMsg.InstanceKey = instanceKey
	qrMsg.ToAddress = addrPort.String()
	qrMsg.Hostname = results[1]
	qrMsg.DriverAddress = results[2]
	qrMsg.Identifier = int32(rand.Int())
	qrMsg.Type = "client_message"
	qrMsg.Retries = 0
	qrMsg.LastResend = time.Now()

	log.Printf("ForwardMessageToQR: %s - %d\n", qrMsg.ToAddress, qrMsg.InstanceKey)

	var qrMsgKey = h.getClientMessageUniqueKey(qrMsg)
	h.forwardedMessages[qrMsgKey] = qrMsg
	h.publishQRMessage(qrMsg)

}
func (h *ClientMessageForwarder) publishQRMessage(msg QRMessage) {
	jsonData, jsonErr := json.Marshal(msg)
	if jsonErr != nil && len(jsonErr.Error()) > 0 {
		log.Printf("publishQRMessage json marshal error: %s\n", jsonErr.Error())
		return
	}

	log.Printf("publishQRMessage: %s - %d\n", msg.ToAddress, msg.InstanceKey)

	err := h.amqpChannel.PublishWithContext(h.context,
		"openspy.master", // exchange
		"qr.message",     // routing key
		false,            // mandatory
		false,            // immediate
		amqp.Publishing{
			ContentType: "text/plain",
			Body:        jsonData,
		})
	if err != nil {
		log.Panicf("Failed to publish message: %s\n", err.Error())
		return
	}
}
func (h *ClientMessageForwarder) OnClientMessageAck(amqpMsg amqp.Delivery) {
	//clear retry logic
	var qrMsg QRMessage
	jsonErr := json.Unmarshal(amqpMsg.Body, &qrMsg)
	if jsonErr != nil && len(jsonErr.Error()) > 0 {
		log.Printf("OnClientMessageAck json marshal error: %s\n", jsonErr.Error())
		return
	}

	qrMsg.ToAddress = qrMsg.FromAddress //handle inversion from qr server

	log.Printf("OnClientMessageAck: %s - %d\n", qrMsg.ToAddress, qrMsg.InstanceKey)

	var qrMsgKey = h.getClientMessageUniqueKey(qrMsg)
	delete(h.forwardedMessages, qrMsgKey)
}

func (h *ClientMessageForwarder) submitResends() {
	for msgKey, value := range h.forwardedMessages {
		if value.Retries >= NUM_PROBE_RETRIES {
			delete(h.forwardedMessages, msgKey)
			continue
		}
		if time.Since(value.LastResend) < time.Duration(2*time.Second) {
			continue
		}
		value.LastResend = time.Now()
		value.Retries = value.Retries + 1
		h.forwardedMessages[msgKey] = value
		h.publishQRMessage(value)

	}
}

func (h *ClientMessageForwarder) getClientMessageUniqueKey(msg QRMessage) string {
	return msg.Hostname + ":" + msg.DriverAddress + ":" + msg.ToAddress + ":" + strconv.Itoa(msg.InstanceKey) + ":" + strconv.Itoa(int(msg.Identifier))
}
